src/TakeoffController.cpp

```diff
@@ -0,0 +1,191 @@
+{
+  // Saving this here to prevent compile warnings. It will be needed in a soon to come revision.
+  ros::NodeHandle nh_ = nh;
+}
+
+// Used to reset and check initial conditions for takeoff
+// Update needs to begin being called shortly after this is called.
+bool TakeoffController::resetForTakeover(const ros::Time& time)
+{
+    if (time < last_update_time_) {
+        ROS_ERROR("Tried to reset TakeoffHandler with time before last update");
+        return false;
+    }
+
+    // Get the current transform (xyz) of the quad
+    geometry_msgs::TransformStamped transform;
+    bool success = transform_wrapper_.getTransformAtTime(transform,
+                                                    "foot1",
+                                                    "map",
+                                                    time,
+                                                    update_timeout_);
+
+    if (!success) {
+        ROS_ERROR("Failed to get current transform in TakeoffController::update");
+        return false;
+    }
+
+    if(transform.transform.translation.z > max_takeoff_start_height_) {
+        ROS_ERROR("Tried to reset the takeoff controller without being on the ground");
+        return false;
+    } else if (state_ != TakeoffState::DONE) {
+        ROS_ERROR("Tried to reset takeoff controller that wasn't in the done state");
+        return false;
+    }
+
+    recorded_hover_throttle_ = 0;
+    throttle_ = 0;
+    state_ = TakeoffState::RAMP;
+    // Mark the last update time as the current time to prevent large throttle spikes
+    last_update_time_ = time;
+    return true;
+}
+
+// Main update
+bool TakeoffController::update(const ros::Time& time,
+                            iarc7_msgs::OrientationThrottleStamped& uav_command)
+{
+    if (time < last_update_time_) {
+        ROS_ERROR("Tried to update TakeoffHandler with time before last update");
+        return false;
+    }
+
+    // Get the current transform (xyz) of the quad
+    geometry_msgs::TransformStamped transform;
+    bool success = transform_wrapper_.getTransformAtTime(transform,
+                                                    "foot1",
+                                                    "map",
+                                                    time,
+                                                    update_timeout_);
+    if (!success) {
+        ROS_ERROR("Failed to get current transform in TakeoffController::update");
+        return false;
+    }
+
+    if(state_ == TakeoffState::RAMP) {
+        if(transform.transform.translation.z > max_takeoff_start_height_) {
+          recorded_hover_throttle_ = throttle_;
+          throttle_ = std::min(throttle_ + takeoff_throttle_bump_, 100.0);
+          state_ = TakeoffState::ASCEND;
+        }
+        else
+        {
+          throttle_ = std::min(throttle_ + ((time - last_update_time_).toSec() * takeoff_throttle_ramp_rate_), 100.0);
+        }
+    }
+    else if(state_ == TakeoffState::ASCEND)
```

Comment:
I disagree, because of the minimum maneuver height.  LLM should be using a velocity controller as soon as we're in the air, but shouldn't allow maneuvers to go all the way down to 0